Lib/fontTools/cffLib/cff2mergePen.py

```diff
@@ -0,0 +1,292 @@
+							  point_type=point_type, pt_index=pt_index,
+							  m_index=m_index, default_type=default_type)
+					][0]
+		super(MergeTypeError, self).__init__(self.error_msg)
+
+
+
+
+def makeRoundNumberFunc(tolerance):
+	if tolerance < 0:
+		raise ValueError("Rounding tolerance must be positive")
+
+	def roundNumber(val):
+		return t2c_round(val, tolerance)
+
+	return roundNumber
+
+
+class CFF2CharStringMergePen(T2CharStringPen):
+	"""Pen to merge Type 2 CharStrings.
+	"""
+	def __init__(self, default_commands,
+				 glyphName, num_masters, master_idx, roundTolerance=0.5):
+		super(
+			CFF2CharStringMergePen,
+			self).__init__(width=None,
+						   glyphSet=None, CFF2=True,
+						   roundTolerance=roundTolerance)
+		self.pt_index = 0
+		self._commands = default_commands
+		self.m_index = master_idx
+		self.num_masters = num_masters
+		self.prev_move_idx = 0
+		self.glyphName = glyphName
+		self.roundNumber = makeRoundNumberFunc(roundTolerance)
+
+	def _p(self, pt):
+		""" Unlike T2CharstringPen, this class stores absolute values.
+		This is to allow the logic in check_and_fix_closepath() to work,
+		where the current or previous absolute point has to be compared to
+		the path start-point.
+		"""
+		self._p0 = pt
+		return list(self._p0)
+
+	def make_flat_curve(self, prev_coords, cur_coords):
+		# Convert line coords to curve coords.
+		dx = self.roundNumber((cur_coords[0] - prev_coords[0])/3.0)
+		dy = self.roundNumber((cur_coords[1] - prev_coords[1])/3.0)
+		new_coords = [prev_coords[0] + dx,
+					  prev_coords[1] + dy,
+					  prev_coords[0] + 2*dx,
+					  prev_coords[1] + 2*dy
+					  ] + cur_coords
+		return new_coords
+
+	def make_curve_coords(self, coords, is_default):
+		# Convert line coords to curve coords.
+		prev_cmd = self._commands[self.pt_index-1]
+		if is_default:
+			new_coords = []
+			for i, cur_coords in enumerate(coords):
+				prev_coords = prev_cmd[1][i]
+				master_coords = self.make_flat_curve(prev_coords[:2],
+													 cur_coords)
+				new_coords.append(master_coords)
+		else:
+			cur_coords = coords
+			prev_coords = prev_cmd[1][-1]
+			new_coords = self.make_flat_curve(prev_coords[:2], cur_coords)
+		return new_coords
+
+	def check_and_fix_flat_curve(self, cmd, point_type, pt_coords):
+		if (point_type == 'rlineto') and (cmd[0] == 'rrcurveto'):
+			is_default = False
+			pt_coords = self.make_curve_coords(pt_coords, is_default)
+			success = True
+		elif (point_type == 'rrcurveto') and (cmd[0] == 'rlineto'):
+			is_default = True
+			expanded_coords = self.make_curve_coords(cmd[1], is_default)
+			cmd[1] = expanded_coords
+			cmd[0] = point_type
+			success = True
+		else:
+			success = False
+		return success, pt_coords
+
+	def check_and_fix_closepath(self, cmd, point_type, pt_coords):
+		""" Some workflows drop a lineto which closes a path.
+		Also, if the last segment is a curve in one master,
+		and a flat curve in another, the flat curve can get
+		converted to a closing lineto, and then dropped.
```

Comment:
I don't think you should elevate a line to a flat curve, as there are infinite ways to do that and there's no way to know which one was the designer's intention.  This just masks a workflow bug, and introduces bugs in the font.  We discussed this in length before.

Reply:
Can you point me to the discussion about the problems with guessing the designers intentions? I agree that converting a line to curve arbitrarily, which I did do this PR, is a bad idea. However, I would have thought that we do have a guide to the designer's intent in the other master designs. We could  use the first master which has a curve as a model for the proportional length of the of the BCP arms. This is mentioned as away to avoid kinks in the 1997 Adobe Tech Note 5091 "Designing Multiple Master Tyepfaces", page 12, on 
https://www.adobe.com/devnet/font.html.

 I'll also need to take a look at where this happens. I think that makeotf and Robofont and Prepolator all convert flat curves to lines. I can fix the latter to have a non-optimizing mode, but if it is possible to adequately fix this problem, I'd argue that it is useful for robustness to handle this case.

Reply:
The starts of the last sentence should have been  'I can fix the first', not 'the latter', i.e. makeotf.

Reply:
Previous discussion is not as heated as I thought it was: https://github.com/googlei18n/cu2qu/issues/55

The way to avoid kinks is to either have same BCP ratios OR same angles.  So, you can't just use BCP ratios...  Also, the BCP ratios are ratios of two BCPs of one point, NOT two BCPs of one curve.  So we end up having to look at neighboring curves.  This is messy.  I don't like to see such hacks.  Definitely not in the name of "robustness".



> I'll also need to take a look at where this happens. I think that makeotf and Robofont and Prepolator all convert flat curves to lines. I can fix the latter to have a non-optimizing mode, but if it is possible to adequately fix this problem, I'd argue that it is useful for robustness to handle this case.

Then we also fix Robofont and Prepolator.  Both have maintainers that frequent this space and can be talked to and fix things properly.  No hacks needed.

Reply:
Fair enough. I will back out flat curve code.